src/core/lib/gpr/tls.h

```diff
@@ -21,52 +21,121 @@
-     gpr_tls_set(&foo, new_value);
+template <typename T, typename = typename std::enable_if<(
+                          std::is_trivially_destructible<T>::value &&
+                          sizeof(T) <= sizeof(void*) &&
+                          alignof(void*) % alignof(T) == 0)>::type>
+class TlsTypeWrapper {
+  static_assert(std::is_trivially_destructible<T>::value &&
+                    sizeof(T) <= sizeof(void*) &&
+                    alignof(void*) % alignof(T) == 0,
+                "unsupported type for TLS");
 
-   Accessing a thread local:
-     current_value = gpr_tls_get(&foo);
+ public:
+  using Type = T;
+};
 
-   ALL functions here may be implemented as macros. */
+}  // namespace grpc_core
 
 #ifdef GPR_STDCPP_TLS
-#include "src/core/lib/gpr/tls_stdcpp.h"
+#define GPR_THREAD_LOCAL(type) \
+  thread_local grpc_core::TlsTypeWrapper<type>::Type
 #endif
 
 #ifdef GPR_GCC_TLS
-#include "src/core/lib/gpr/tls_gcc.h"
+#define GPR_THREAD_LOCAL(type) __thread grpc_core::TlsTypeWrapper<type>::Type
 #endif
 
 #ifdef GPR_MSVC_TLS
-#include "src/core/lib/gpr/tls_msvc.h"
+#define GPR_THREAD_LOCAL(type) \
+  __declspec(thread) grpc_core::TlsTypeWrapper<type>::Type
 #endif
 
 #ifdef GPR_PTHREAD_TLS
-#include "src/core/lib/gpr/tls_pthread.h"
+#include <grpc/support/log.h> /* for GPR_ASSERT */
+#include <pthread.h>
+
+namespace grpc_core {
+
+template <typename T>
+class pthreadTlsWrapper : TlsTypeWrapper<T> {
+ public:
+  pthreadTlsWrapper() = default;
+  pthreadTlsWrapper(const pthreadTlsWrapper&) = delete;
+  pthreadTlsWrapper& operator=(const pthreadTlsWrapper&) = delete;
+
+  void Init() { GPR_ASSERT(0 == pthread_key_create(&key_, nullptr)); }
+
+  void Destroy() { GPR_ASSERT(0 == pthread_key_delete(key_)); }
+
+  operator T() const { return Cast<T>(pthread_getspecific(key_)); }
+
+  T operator=(T t) {
+    GPR_ASSERT(0 == pthread_setspecific(key_, Cast<T>(t)));
+    return t;
+  }
+
+ private:
+  // TODO(C++17): Replace these helpers with constexpr if statements inline.
+  template <typename V>
+  static typename std::enable_if<std::is_pointer<V>::value, V>::type Cast(
+      void* object) {
+    return static_cast<V>(object);
+  }
+
+  template <typename V>
+  static typename std::enable_if<
+      !std::is_pointer<V>::value && std::is_integral<V>::value, V>::type
+  Cast(void* object) {
+    return reinterpret_cast<uintptr_t>(object);
+  }
+
+  template <typename V>
+  static void* Cast(
+      typename std::enable_if<std::is_pointer<V>::value, V>::type t) {
+    return t;
+  }
+
+  template <typename V>
+  static void* Cast(typename std::enable_if<!std::is_pointer<V>::value &&
+                                                std::is_integral<V>::value,
+                                            V>::type t) {
+    return reinterpret_cast<void*>(uintptr_t(t));
+  }
+
+  pthread_key_t key_;
+};
+
+}  // namespace grpc_core
+
+#define GPR_THREAD_LOCAL(type) grpc_core::pthreadTlsWrapper<type>
+
+#define gpr_tls_init(tls) (tls).Init()
```

Comment:
If we are making this a C++ API anyway, why not just change callers to directly use the C++ API, rather than wrapping it in this C-like interface?  We could presumably provide the same C++ API for all thread-local implementations.

Moving to gprpp makes sense for C++ code, but this C-style API would probably want to stay in gpr if we keep it.  But if there are no callers of this API outside of core, then we don't need the C-style API to begin with.

Reply:
Because there's no wrapper class *being instantiated* in the not-pthread case. What would there be a method on?

Reply:
Regarding the location: is that determined by the API *style* or by its language compatibility? This commit changes this header such that it won't compile in C, but I defer to you on what the heuristic should be.

Reply:
It's generally about the API style.  We can compile code written in C++ no matter what directory it's in.

If the API here is still a C-style API, then I guess keeping it in the gpr directory is fine.

Reply:
Ok, I will move it back. Anything else I should address?

Reply:
Moved back to gpr.